File: app/concepts/matestack/ui/core/action/action.rb
module Matestack::Ui::Core::Action
  class Action < Matestack::Ui::Core::Component::Dynamic
    optional :path, :success, :failure, :notify, :confirm, 
      method: { as: :action_method }, params: { as: :action_params }

    def vuejs_component_name
      'matestack-ui-core-action'
    end

    def setup
      @component_config[:action_path] = action_path
      @component_config[:method] = action_method
      @component_config[:success] = success
      unless success.nil?
        unless success[:transition].nil?
          @component_config[:success][:transition][:path] = transition_path success
        end
        unless success[:redirect].nil?
          @component_config[:success][:redirect][:path] = redirect_path success
        end
      end
      @component_config[:failure] = failure
      unless failure.nil?
        unless failure[:transition].nil?
          @component_config[:failure][:transition][:path] = transition_path failure
        end
        unless failure[:redirect].nil?
          @component_config[:failure][:redirect][:path] = redirect_path failure
        end
      end
      if notify.nil?
        @component_config[:notify] = true
      end
      if @component_config[:confirm] = confirm
        @component_config[:confirm_text] = confirm.try(:[], :text) || "Are you sure?"
      end
    end

    def action_path
      begin
        if path.is_a?(Symbol)
          return ::Rails.application.routes.url_helpers.send(path, action_params)
        else
          return path
        end
      rescue
        raise "Action path not found"
      end
    end

    def transition_path callback_options
      begin
        if callback_options[:transition][:path].is_a?(Symbol)
          return ::Rails.application.routes.url_helpers.send(
            callback_options[:transition][:path],
            callback_options[:transition][:params]
          )
        else
          return callback_options[:transition][:path]
        end
      rescue
        raise "Transition path not found"
      end
    end

    def redirect_path callback_options
      begin
        if callback_options[:redirect][:path].is_a?(Symbol)
          return ::Rails.application.routes.url_helpers.send(
            callback_options[:redirect][:path],
            callback_options[:redirect][:params]
          )
        else
          return callback_options[:redirect][:path]
        end
      rescue
        raise "Redirect path not found"
      end
    end

  end
end

File: app/concepts/matestack/ui/core/action/action.js
import Vue from 'vue/dist/vue.esm'
import Vuex from 'vuex'
import axios from 'axios'

import matestackEventHub from '../js/event-hub'
import componentMixin from '../component/component'

const componentDef = {
  mixins: [componentMixin],
  data: function(){
    return {}
  },
  methods: {
    perform: function(){
      const self = this
      if (
        (self.componentConfig["confirm"] == undefined) || confirm(self.componentConfig["confirm_text"])
      )
      {
        if (self.componentConfig["emit"] != undefined) {
          matestackEventHub.$emit(self.componentConfig["emit"]);
        }
        if (self.componentConfig["delay"] != undefined) {
          setTimeout(function () {
            self.sendRequest()
          }, parseInt(self.componentConfig["delay"]));
        } else {
          this.sendRequest()
        }
      }
    },
    sendRequest: function(){
      const self = this
      axios({
          method: self.componentConfig["method"],
          url: self.componentConfig["action_path"],
          data: self.componentConfig["data"],
          headers: {
            'X-CSRF-Token': document.getElementsByName("csrf-token")[0].getAttribute('content')
          }
        }
      )
      .then(function(response){
        if (self.componentConfig["success"] != undefined && self.componentConfig["success"]["emit"] != undefined) {
          matestackEventHub.$emit(self.componentConfig["success"]["emit"], response.data);
        }

        // transition handling
        if (self.componentConfig["success"] != undefined
          && self.componentConfig["success"]["transition"] != undefined
          && (
            self.componentConfig["success"]["transition"]["follow_response"] == undefined
            ||
            self.componentConfig["success"]["transition"]["follow_response"] === false
          )
          && self.$store != undefined
        ) {
          let path = self.componentConfig["success"]["transition"]["path"]
          self.$store.dispatch('navigateTo', {url: path, backwards: false})
          return;
        }
        if (self.componentConfig["success"] != undefined
          && self.componentConfig["success"]["transition"] != undefined
          && self.componentConfig["success"]["transition"]["follow_response"] === true
          && self.$store != undefined
        ) {
          let path = response.data["transition_to"] || response.request.responseURL
          self.$store.dispatch('navigateTo', {url: path, backwards: false})
          return;
        }
        // redirect handling
        if (self.componentConfig["success"] != undefined
          && self.componentConfig["success"]["redirect"] != undefined
          && (
            self.componentConfig["success"]["redirect"]["follow_response"] == undefined
            ||
            self.componentConfig["success"]["redirect"]["follow_response"] === false
          )
          && self.$store != undefined
        ) {
          let path = self.componentConfig["success"]["redirect"]["path"]
          window.location.href = path
          return;
        }
        if (self.componentConfig["success"] != undefined
          && self.componentConfig["success"]["redirect"] != undefined
          && self.componentConfig["success"]["redirect"]["follow_response"] === true
          && self.$store != undefined
        ) {
          let path = response.data["redirect_to"] || response.request.responseURL
          window.location.href = path
          return;
        }
      })
      .catch(function(error){
        if (self.componentConfig["failure"] != undefined && self.componentConfig["failure"]["emit"] != undefined) {
          matestackEventHub.$emit(self.componentConfig["failure"]["emit"], error.response.data);
        }
        // transition handling
        if (self.componentConfig["failure"] != undefined
          && self.componentConfig["failure"]["transition"] != undefined
          && (
            self.componentConfig["failure"]["transition"]["follow_response"] == undefined
            ||
            self.componentConfig["failure"]["transition"]["follow_response"] === false
          )
          && self.$store != undefined
        ) {
          let path = self.componentConfig["failure"]["transition"]["path"]
          self.$store.dispatch('navigateTo', {url: path, backwards: false})
          return;
        }
        if (self.componentConfig["failure"] != undefined
          && self.componentConfig["failure"]["transition"] != undefined
          && self.componentConfig["failure"]["transition"]["follow_response"] === true
          && self.$store != undefined
        ) {
          let path = error.response.data["transition_to"] || response.request.responseURL
          self.$store.dispatch('navigateTo', {url: path, backwards: false})
          return;
        }
        // redirect handling
        if (self.componentConfig["failure"] != undefined
          && self.componentConfig["failure"]["redirect"] != undefined
          && (
            self.componentConfig["failure"]["redirect"]["follow_response"] == undefined
            ||
            self.componentConfig["failure"]["redirect"]["follow_response"] === false
          )
          && self.$store != undefined
        ) {
          let path = self.componentConfig["failure"]["redirect"]["path"]
          window.location.href = path
          return;
        }
        if (self.componentConfig["failure"] != undefined
          && self.componentConfig["failure"]["redirect"] != undefined
          && self.componentConfig["failure"]["redirect"]["follow_response"] === true
          && self.$store != undefined
        ) {
          let path = error.response.data["redirect_to"] || response.request.responseURL
          window.location.href = path
          return;
        }
      })
    }
  }
}

let component = Vue.component('matestack-ui-core-action', componentDef)

export default componentDef
